### PR TITLE
Git practices

### PR DESCRIPTION
**Pull request by maurolepore**
Dear reviewers, 

This PR adds a new document "git.md" that documents git practices. You may preview the evolving document [here](https://github.com/2DegreesInvesting/practices/blob/git-practices/git.md). 

Please add your review. Go to "Files changed", click on "Review changes", and either "Comment", "Approve", or "Request changes" (see each option's description).

Because @MonikaFu and I decided to agree by consensus, to merge this PR will have to adapt until we all feel comfortable to  "Approve".

- [ ] Alex <alex@2degrees-investing.org>
- [ ] Andreas <andy@2degrees-investing.org>
- [ ] CJ <cj@2degrees-investing.org>
- [ ] Clare <clare@2degrees-investing.org>
- [ ] Constanze <constanze@2degrees-investing.org>
- [ ] Jackson <jackson@2degrees-investing.org>
- [ ] Jacob Kastl <jacob@2degrees-investing.org>
- [x] Mauro <mauro@2degrees-investing.org>
- [ ] Taylor <taylor@2degrees-investing.org>
- [ ] Catarina <catarina@2degrees-investing.org>
- [ ] Fiona <fiona@2degrees-investing.org>.
- [ ] Frederick <frederick@2degrees-investing.org>
- [x] Monika <monika@2degrees-investing.org>

### CODE REVIEW

File: git.md [hunk 1]
```diff
@@ -1,2 +1,130 @@
-# Git
+These practices:
```
Comment (maurolepore) [inline git.md:1]: @MonikaFu, @jdhoffa, and @jacobvjk, here is the latest draft. Can you please do a quick review before we ask for everyone else's review?

At the end you'll see a "Discussion" section. I plan to move that text out from here and into the "Discussions" tab of this repo.

File: git.md [hunk 2]
```diff
@@ -1,2 +1,130 @@
...
+code behaves as you expect, and how someone else may reproduce your results.
+
+## 6. Structure each commit message like an email
+
+Like an email, commit messages have a subject and an optional body. The subject
+should start with a line summarizing your changes (aim for 50 characters or
+less). Use the imperative, present tense -- for example, "Fix typo",
+not "Fixed typo" or "Fixes typo". The message body, if present,
+should follow a blank line; wrap around 70 characters; and explain the
+motivation for the change ("why", not "what" or "how").
```
Comment (jdhoffa) [inline git.md:68]: I didn't know that the focus of the body is why, and not what or how! That's cool.

**Review by jdhoffa — APPROVED** [review]
This looks good! Succinctly covers all the important topics.

File: git.md (hunk 2 above)
Comment (maurolepore) [reply git.md:68]: Yeah, that's cool. I guess the reason is that the intent is the only thing that the changes themselves don't capture, so explaining "why" is most important. Summarizing "what" or "how" might still be nice, but the effortful reader can get it from the change history.

File: git.md [hunk 3]
```diff
@@ -1,2 +1,127 @@
...
+* <https://github.com/2DegreesInvesting/coding-helpdesk#coding-helpdesk>.
+* <https://github.com/git-guides/>.
+* <https://coderefinery.github.io/git-intro/>.
+* <https://bit.ly/book-git-in-practice (Part 4: Git best practices)>.  
+* <https://dev.to/lydiahallie/cs-visualized-useful-git-commands-37p1>.
+
+## Discuss
+
+* Jackson: Always include files like LICENSE, README.md, and consider
+.gitignore and CONTRIBUTING.md.
```
Comment (maurolepore) [inline git.md:123]: @jdhoffa, this seems a bit out of scope for the Git theme (except maybe .gitignore). Should this move to another theme, maybe "Languagge agnostic practices"?

File: git.md [hunk 4]
```diff
@@ -1,2 +1,127 @@
...
+* <https://coderefinery.github.io/git-intro/>.
+* <https://bit.ly/book-git-in-practice (Part 4: Git best practices)>.  
+* <https://dev.to/lydiahallie/cs-visualized-useful-git-commands-37p1>.
+
+## Discuss
+
+* Jackson: Always include files like LICENSE, README.md, and consider
+.gitignore and CONTRIBUTING.md.  
+* Taylor: Develop a consensus for types of changes, e.g. "amend" (something
+I thought I had already done).
```
Comment (maurolepore) [inline git.md:125]: @tposey28, Can you expand a bit on this idea, maybe with a few examples? Also which subsection would you fit it, maybe workflow?

File: git.md (hunk 1 above)
Comment (maurolepore) [reply git.md:1]: @MonikaFu and @jacobvjk, just a friendly reminder of my request for your quick review before I ask for everyone else's review. Would you be able to share you opinion by the end of this week?

File: git.md (hunk 4 above)
Comment (tposey28) [reply git.md:124]: I think it goes in workflow.
My intent with this was that a developer who wants to amend a commit can do one of two things:
1. They can literally amend their latest commit, using the `-amend` command. It can be used to change the contents or the commit message, but is best only used if the commit hasn't been pushed yet (if it has then a merge conflict will occur, which is a bit confusing, plus at this point the commit has been "shared" in a sense).
2. If the commit is no longer the latest (or we want to avoid using the `-amend` flag because it's technically rewriting history) then I was suggesting that one could literally use a commit message like the following: "Amend: Added behavior to script". This isn't something I've ever heard of in practice, it's just an idea I had that could make following the flow easier in a PR.

Best case scenario you never need to amend ;)

**Review by tposey28 — APPROVED** [review]

File: git.md (hunk 4 above)
Comment (tposey28) [reply git.md:124]: https://careerkarma.com/blog/git-commit-amend/

Comment (jdhoffa) [reply git.md:124]: This is an interesting idea, but it seems a bit niche/ not frequently used to be included as a "best practice". Maybe we could add an addendum with "nice-to-haves" or something like that? 
@maurolepore @MonikaFu @cjyetman @everyone thoughts?

**Review by jacobvjk — APPROVED** [review]
I'm happy with this and support @jdhoffa idea to use a nice-to-have section for more niche applications

File: git.md [hunk 5]
```diff
@@ -1,2 +1,125 @@
...
+not "Fixed typo" or "Fixes typo". The message body, if present,
+should follow a blank line; wrap around 70 characters; and explain the
+motivation for the change ("why", not "what" or "how").
+
+See also:
+
+* [Example of a commit message](https://github.com/2DegreesInvesting/resources/issues/74).
+* <https://r-pkgs.org/git.html#commit-best-practices.>
+
+## 7. Agree on a workflow
```
Comment (MonikaFu) [inline git.md:73]: this should come with a caveat that we should actually agree on a workflow I think. As in - these are OUR best practices so shouldn't we just say: structure your work according to XYZ workflow?

File: git.md (hunk 4 above)
Comment (MonikaFu) [reply git.md:124]: @jdhoffa I like the idea of nice-to-have's and what @tposey28 mentions sounds like an example of a good commit message also so maybe we could use it like that as well.

Comment (maurolepore) [reply git.md:127]: How about writing something like this?:

```
## Agree on a workflow

GitHub workflow

... bla bla

Git workflow

... bla bla

A team may agree on custom conversions beyond the "vanilla" workflows listed above. 
For example, you may agree to prefix a commit title with "WIP", "FIXME", "Amend: " to
communicate with your team succinctly via a vocabulary you all share.
```

File: git.md (hunk 5 above)
Comment (maurolepore) [reply git.md:73]: I'm a little conflicted. 

On the one hand I thing this doc should be opinionated. This supports @MonikaFu's idea of committing to one way of doing things, and suggests that @tposey28's "nice to have's may need to find another place. So it is now the time to take the discomfort of excluding some good ideas in the interest of keeping the minimum we can all live with.

And on the other hand, I know different teams and projects may have different needs, suggesting this document should be a flexible. I believe it's enough to say "agree on a workflow" as long as it's clearly documented. See for example <https://github.com/github/gitignore#contributing-workflow> -- it wouldn't be hard for a new comer to fit in.

File: git.md (hunk 4 above)
Comment (MonikaFu) [reply git.md:127]: @maurolepore sounds great to me

File: git.md (hunk 5 above)
Comment (MonikaFu) [reply git.md:73]: @maurolepore I see what you mean. Good point. At the same time I feel that for people who are less familiar with GitHub choosing a workflow themselves might be intimidating. My proposal taking into account your comment would be to agree on a 'default' workflow with an indication that teams are free to use a different workflow per project when they see a benefit. What do you think?

Comment (maurolepore) [reply git.md:73]: Sounds like an awesome idea -- having a "default" best practice when the number of practices considered "best" is greater than one :)

File: git.md [hunk 6]
```diff
@@ -1,2 +1,139 @@
...
+(or `-f`) when pushing to shared repositories.
+
+See also:
+
+* [Cleaning _unshared_ history: Squashing all commits in a branch](https://youtu.be/08dhy3Zoob4).
+* [Cleaning _unshared_ history with `git rebase --interactive`](https://youtu.be/cMI8p1XhMzA).
+* [Cleaning _unshared_ history with `git commit --amend`](https://youtu.be/539pfVfr7OI).
+* [Cleaning shared history with `git revert`](https://youtu.be/A8Ld6iDqc3w).
+* [Git playlist](https://www.youtube.com/playlist?list=PLvgdJdJDL-AOHkwiaMvYhPKVjiD9vzZIo).
+
```
Comment (maurolepore) [inline git.md:29]: FYI I added links to short videos about the commands we mention in this section.

**Review by Clare2D — APPROVED** [review]
👍

**Review by MonikaFu — APPROVED** [review]
LGTM

**Review by FrederickFa — APPROVED** [review]

**Comment by andybue** [comment]
Thanks

**Review by andybue — APPROVED** [review]

File: git.md [hunk 7]
```diff
@@ -1,2 +1,139 @@
...
+A team may agree on custom conventions beyond the "vanilla" workflows listed
+above.  For example, you may agree to prefix a commit-title with "WIP",
+"FIXME", "Amend: " or to whatever to communicate with your team succinctly
+via a vocabulary you all share.
+
+See also: 
+
+* [Comparing workflows](https://www.atlassian.com/git/tutorials/comparing-workflows).
+
```
Comment (AlexAxthelm) [inline git.md:114]: I think that we should also include a section on the value of local and global `gitignore`s. Local gitignores should call out what should be ignored for that project (for example, data files or rendered outputs), while global gitignores should be the files that each developer should ignore in _every_ project (`.DS_Store`, vim swap files, anything that is specific to my particular machine or text editor configuration)

File: git.md [hunk 8]
```diff
@@ -1,2 +1,139 @@
...
+the goal of one commit requires changes in two files, that one commit should
+include changes to both files. Similarly, if the goal of a feature requires
+two kinds of changes, that feature should likely include two commits.
+
+## 5. Explain how to test the effect of the changes you made
+
+Testing your code is important, particularly when sharing it with
+others. Automated tests are the best but not the only way to test the effects
+of the changes you made. However you did it, explain how you ensured that the
+code behaves as you expect, and how someone else may reproduce your results.
```
Comment (AlexAxthelm) [inline git.md:65]: The instructions for testing should be included in either code or text files, not in git commit messages. commit messages are signposts to help other developers find the state of code at a point in time, not a forum for instruction or discussion.

File: git.md [hunk 9]
```diff
@@ -1,2 +1,139 @@
...
+
+Choose the tools that help you work more comfortably, more productively,
+and with fewer errors.
+
+See also:
+
+* [Install a Git client](https://happygitwithr.com/git-client.html)
+* [oh-my-zsh: Unleash your terminal like never before](https://ohmyz.sh/).
+* [gh: Take GitHub to the command line](https://cli.github.com/).
+* [Git from RStudio](https://rstudio.com/resources/webinars/managing-part-2-github-and-rstudio/).
```
Comment (cjyetman) [inline git.md:125]: These are all so much about personal preference that it seems antithetical to a guidelines document that is/should be making a few hard "rules" about how we work together. Sharing ideas about tools and elective practices is great, but guidelines are specifically about the things that are *not* elective, right?

File: git.md (hunk 7 above)
Comment (maurolepore) [reply git.md:114]: Thanks for pointing that out. 

@jdhoffa also pushed for .gitignore files and now the sectoin _2. Keep commits focused_ inlcudes "You can avoid sharing unwanted changes via .gitignore." 

Your suggestion would fit nicely here:

>  You can avoid sharing unwanted changes via .gitignore. Local gitignores should call out what should be ignored for that project (for example, data files or rendered outputs), while global gitignores should be the files that each developer should ignore in every project (.DS_Store, vim swap files, anything that is specific to my particular machine or text editor configuration)

To keep this document short, this document focuses only on _what_ to do and _why_; discussions about _how_ to do it are referenced via links ([.gitignore templates](.gitignore templates)). @jacobvjk and I discussed this restriction and figured we need to draw the boundary somewhere. But I realize this decision paves the road for documenting other themes so we likely need to engage @MonikaFu and the leaders of other themes like you.

What do you think? Is this a reasonable trade-off? Should we continue in the tab "Discussions"?

**Review by cjyetman — CHANGES_REQUESTED** [review]

File: git.md (hunk 8 above)
Comment (maurolepore) [reply git.md]: Thanks!

After 9dc246d the doc reads 
> When automated tests are excluded, the PR's first comment should
explain how you ensured that the code behaves as you expect,
and how someone else may reproduce your results.

The key phrase here is "the PR's first comment". While this is in line with your suggestion that the comment should not be in the git commit message, it is not exactly what you suggested -- that the comment should be in "either code or text file". Are you okay with this? If not, please suggest the bit of text you would propose (or expand on your idea as I didn't fully understand your suggestion about explaining the testing approach in text files).

File: git.md [hunk 10]
```diff
@@ -1,2 +1,139 @@
...
+
+* Shared commits should not be changed.
+* Changes in each shared commit should be closely related.
+
+## 1. Preserve shared history
+
+Shared commits should never change. Some Git commands preserve the history, but
+others don't. For example, `git commit --amend` and `git rebase --interactive`
+alter the git history; they help to overwrite  a messy commit history with
+a cleaner one; you should use them only before sharing your work. To cleanup
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md:17]: I'm thinking of including a glossary. That way the main text could keep short by using a succinct vocabulary we define elsewhere.

File: git.md [hunk 11]
```diff
@@ -1,2 +1,139 @@
...
+via a vocabulary you all share.
+
+See also: 
+
+* [Comparing workflows](https://www.atlassian.com/git/tutorials/comparing-workflows).
+
+## 8. Use tools to be more productive
+
+Choose the tools that help you work more comfortably, more productively,
+and with fewer errors.
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md:119]: Yeah. Also, it would be out of scope -- if we formalize that this document should focus on _what_ and _why_.

File: git.md (hunk 9 above)
Comment (maurolepore) [reply git.md:126]: Right -- with the caveat that this is aspirational as people are likely to face situations that make strict applicatoin impossible and may need to make informed trade-offs.

File: git.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.md:17]: that sounds excellent to me.... I guess it's not possible in markdown (?) but it would be cool if you could hover over the word to get its definition

File: git.md (hunk 7 above)
Comment (AlexAxthelm) [reply git.md:115]: I fall pretty solidly into the mindset that each person working with the repo is responsible for keeping their contributions clean, not for cleaning up after other people. If editor-specific files are added to the `.gitignore`, then it's easy for that to grow to include a lot of unnecessary filetypes, especially if they are there after people move along. And if we put user-specific files into the local ignore, then we have to put it in there for _every_ repo, which also sounds like a worse option than coaching people to build their own gloabl `.gitignore` and move it from machine to machine with them, like they would any other configuration files.

Comment (cjyetman) [reply git.md:115]: > I fall pretty solidly into the mindset that each person working with the repo is responsible for keeping their contributions clean, not for cleaning up after other people.

Then we don't need `.gitignore` (global or repo-specific) at all... everyone should just do their commits properly and not include stuff that they shouldn't. Obviously, I'm being a bit sarcastic.

> If editor-specific files are added to the .gitignore, then it's easy for that to grow to include a lot of unnecessary filetypes

Hypothetically I agree with this, but I'm willing to bet that if we searched the entire git history of every repo 2dii has ever had that we would find no more than 20 file extensions or filenames of hidden files or other stuff that one would be likely to put in `.gitignore`

**Review by looreen — APPROVED** [review]

File: git.md (hunk 8 above)
Comment (maurolepore) [reply git.md]: @AlexAxthelm while you are here, please confirm you are okay with this or suggest what you would like to read instead.

File: git.md [hunk 12]
```diff
@@ -1,2 +1,139 @@
-# Git
+These practices:
+
+* are aspirational but when they can't be followed it's nice to explain why;
+* document heuristics, as detailed instructions may never cover all scenarios;
+* apply to shared commits (maybe after cleaning commits locally).
+
+## tl;dr
+
+* Shared commits should not be changed.
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md]: f592af0 defines "shared commit" as a footnote.

File: git.md [hunk 13]
```diff
@@ -1,2 +1,139 @@
...
+* Changes in each shared commit should be closely related.
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md:10]: Section "2. Keep commits focused" gives a tiny example to try explain what "closely related" means. Are you okay with this or do you suggest a different example or another complementary example? If so, I'd appreciate you spell it out (I am now a bit low in brain power to create one :-):

> A commit should focus on related changes only. For example, fixing two different bugs should produce two separate commits.

File: git.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md:21]: >  "sharing" also feels ambiguous to me... can we be more precise with this, like "pushed to origin" or something?

Thanks! The definition of "shared commit" now reads "for example, because you pushed it to a remote repository that
someone else has permission privileges to access".

> it would be cool if you could hover over the word to get its definition

The closest I could think of is a footnote, which I now prefer over the glossary because in the source code we can keep the definition close to the term being defined, and in the rendered output the term and its definition are just one click (not hover) away.

File: git.md [hunk 14]
```diff
@@ -1,2 +1,139 @@
...
+* A new release is created by tagging a particular commit on the main branch.
+
+Gitflow workflow:
+
+* Like the GitHub workflow but uses two long-lived branches: "main"
+and "develop". The main branch stores the official release history,
+and the develop branch serves as an integration branch for features.
+
+A team may agree on custom conventions beyond the "vanilla" workflows listed
+above.  For example, you may agree to prefix a commit-title with "WIP",
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md]: Thanks for giving me access to the reader's mind. 364c648 removes the reference to the "WIP" prefix to avoid distracting the reader with a thought of whether or not that is even possible. 

(BTW, I use WIP a lot locally but -- as you suggest -- I cleanup before I share that work -- so that's irrelevant to this document which focuses on shared work).

File: git.md [hunk 15]
```diff
@@ -1,2 +1,139 @@
...
+
+This workflow is 2DII's default. If a project uses a different
+workflow, document prominently (e.g. in README) what that workflow is
+([example](https://github.com/github/gitignore#contributing-workflow)).
+
+* Uses a central repository. 
+* Uses a single long-lived branch: "main"; it never contains broken code.
+* Developers work on "feature" branches (never on the "main" branch), share
+them via pull requests, and peers review the changes before they become a
+part of the main codebase.
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md]: Thanks for noticing that. After d7cf0e5 this doc refers to 'central repository'. I considered upstream but that might be confusing because the remote name might be anything (e.g. origin).

File: git.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.md]: Thanks! I rewrote to focus on systems. I still mention tools but only as an example of how to do something in a systematic way.

7b38cb6

File: git.md (hunk 9 above)
Comment (maurolepore) [reply git.md]: While this document avoids details on _how_ to do things, it does point to external resources. These entries, in particular, respond to a request by @MonikaFu and @catarinabrg -- who liked some kind of optional guidance which may save some folks considerable time and frustration. 

In 4134e8e I start the document with this:

> This document describes git practices we agree by consensus. It focuses on _what_ to do and _why_, and only recommends optional resources about _how_ to do it.

This clarifies that the linked resources are optional (i.e. an opinion rather than a guideline). @cjyetman, is this a reasonable trade-off between your comment and the request by @MonikaFu and @catarinabrg?

Also, this comment addresses previous discussion with @AlexAxthelm -- where I said this document focuses on _what_ and _why_ but that wasn't explicit anywhere in the document -- until now.

File: git.md (hunk 7 above)
Comment (cjyetman) [reply git.md:133]: I don't know of any links that specifically address the debate of using global .gitignore versus repo specific .gitignore versus not using either, but I agree that whether or not and how one uses .gitignore is an implementation detail. The guideline is: do not commit unnecessary files into the repo. The advice is: consider using global and/or repo .gitignore files to help you not do that.

https://docs.github.com/en/github/using-git/ignoring-files
https://gist.github.com/octocat/9257657
https://github.com/github/gitignore/blob/master/R.gitignore

Comment (maurolepore) [reply git.md:133]: Note to myself: Maybe link the word .gitignore to https://git-scm.com/docs/gitignore

File: git.Rmd [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+to be focused; they are easier to review, and easier to merge because they
+offer fewer opportunities for conflicts.
+
+## 4. The work you share should be complete
+
+Each commit and feature should do what it promises, completely; and reverting
+it should undo what that commit or feature meant to do, also completely. If
+the goal of one commit requires changes in two files, that one commit should
+include changes to both files. Similarly, if the goal of a feature requires
+two kinds of changes, that feature should likely include two commits.
```
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd:66]: Maybe we can provide an example for the reader of a situation of where a commit does not "do what it promises, completely"? I personally can't think of what that would look like.

File: git.Rmd [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+
+* [Comparing workflows](https://www.atlassian.com/git/tutorials/comparing-workflows).
+
+## 8. Use tools to be more productive
+
+Rather thank relying on memory or will power, rely on good systems. Good systems
+such as git [clients](https://happygitwithr.com/git-client.html),
+[aliases](https://git-scm.com/book/en/v2/Git-Basics-Git-Aliases), and
+[hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) can help you
+work more comfortably, more productively, and with fewer errors.
```
[the remark this answers is not in the document]
Comment (jdhoffa) [reply git.Rmd:130]: I believe git client = GUI in this context.

File: git.Rmd [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+
+* Shared commits^[A _shared_ commit is a commit that someone other than you
+might access, for example, because you pushed it to a remote repository that
+someone else has permission privileges to access.] should not be changed.
+* Changes in each shared commit should be closely related.
+
+## 1. Preserve shared history
+
+Shared commits should never change. Some Git commands preserve the history, but
+others don't. For example, `git commit --amend` and `git rebase --interactive`
```
[the remark this answers is not in the document]
Comment (jdhoffa) [reply git.Rmd:23]: I think in feature branches this would actually be even more dangerous, as someone is more likely to potentially try to commit/ pull from your working branch. If there is a situation in which someone is likely to have pulled your work and is committing on top of it, you should aim to not alter the history. 

It doesn't mean that you can't clean things up before they are merged (ie you can of course squash/ rebase when the PR is approved/ merged), and it definitely doesn't mean that everything that gets pushed upstream/ origin needs to be super pristine, it just means that the act of pushing sorta means you can't just delete that commit like it never existed. 

Tools like `git revert` allow you to revert a commit by adding the inverse commit on top of it, so as to not alter the history, that would be my recommendation instead of git rebase.

File: git.Rmd (hunk 17 above)
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd:130]: good point, I did not realize that

[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd:130]: consider this resolved.... looks like I don't have permission to do that?

File: git.Rmd (hunk 18 above)
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd:23]: I tried my best to rewrite more clearly. The idea of not changing shared-commits applies both to workflows that do and don't use forks. I'm finding it hard to express it succinctly and completely. Maybe I need to complement the text with a demo.

File: git.Rmd [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+A commit should focus on related changes only. For example, fixing two
+different bugs should produce two separate commits. Small commits make it
+easier for other developers to understand the changes and roll them back if
+something went wrong. You can avoid sharing unwanted changes via .gitignore. If
+you can't succinctly describe what the commit is doing in a couple of words,
+then it is likely too big.
+
+See also:
+
+* [The right size of a PR](https://github.com/2DegreesInvesting/practices/discussions/3).
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd:46]: The relationship I see is that pull requests are made up of multiple, focused commits. I think it's like talking about atoms and pointing to information about molecues.

File: git.Rmd [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+* [.gitignore templates](https://github.com/github/gitignore).
+* Undoing things with Git ([video](https://youtu.be/dZOfEF19yDk),
+[lesson](https://coderefinery.github.io/git-intro/05-undoing)).
+
+## 3. Commit frequently
+
+Committing frequently keeps your commits small. Small commits are more likely
+to be focused; they are easier to review, and easier to merge because they
+offer fewer opportunities for conflicts.
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd:55]: Yeah, I agree. I merged the two sections as they are essentially the same.

File: git.Rmd (hunk 16 above)
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd:63]: I think the issue is clearer if we exaggerate your  example. If two commits have the title "Fix all typos" then none does the job completely. That's fine locally but one should thrive to squash those two incomplete commits and instead share only one that really fixes all typos. Do you still think this section is needless?

File: README.md [hunk 21]
```diff
@@ -8,5 +8,11 @@ that defines a "best practice". Anecdotes and personal experience have
 value, but we prefer evidence from research, books, and large and mature
 communities of practice.
 
+We have agreed to this practices by consensus. We document _what_ to do and
+_why_, and only recommend optional resources about _how_ to do it. These
+practices are aspirational but when they can't be followed it's nice to explain
+why. They document heuristics, as detailed instructions may never cover all
+scenarios.
+
```
Comment (maurolepore) [inline README.md:16]: @MonikaFu, I elevated this paragraph to README because it applies to all themes. Do you agree?

File: git.Rmd (hunk 16 above)
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd]: That makes more sense. But it makes more sense in the context of looking back at commits you've already made and saying, "shucks, I should've just put those last few into one commit", versus giving a good rule of thumb to go by when/before making the commits in the first place. I guess you would suggest that if you've noticed something like that, then you should modify your commits/history locally so that they're combined into a single commit before pushing it to the shared repo? I'm not sure that's a great idea... or at least not something we need to encourage people, in particular those who are not super fluent in git (myself included), to do. And ultimately what's the value of that anyway? We push commits to a feature branch, and then ultimately all of those commits get squash-merged into master and the branch is deleted, so.... that seems like more effort than it's worth.

[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd]: The paragraph now includes:
>  For example, if two commits in the same feature branch promise to “Fix all typos”, each
> commit is likely incomplete; they should have been squashed into a single commit before 
> sharing them.

Which rule of thumb would you propose? 

In my opinion, its' okay to encourage people to rewrite their commit history locally -- as much as I encourage rewriting a sentence after drafting it. Few people if any write clear commits or sentences the first try. 

The value of thriving for complete commits in a working feature is that -- even if the feature will be squashed -- each commit can be reverted easily.  It also allows the reviewer to understand the sequence of changes by navigating commits one by one with `n` and `p` under the [commits tab](https://github.com/2DegreesInvesting/practices/pull/7/commits/), rather than the [files changed tab](https://github.com/2DegreesInvesting/practices/pull/7/files).  I find that feature quite usefule in a PR like this one where you may want to see how your comments translated to commits -- for example starting at https://github.com/2DegreesInvesting/practices/pull/7/commits/3897bac0bff0e915919bc9f9afaa23b5a5f6b199 and hitting `n` until the end.

reverting it should undo what that commit or feature meant to do


https://github.com/2DegreesInvesting/practices/pull/7/commits/38b730cb0ab2527e2a79d349f1c5a9d7f733e1bf

File: git.Rmd [hunk 22]
```diff
@@ -0,0 +1,151 @@
+---
+output: github_document
+---
+
+This document describes git practices we agree by consensus. It focuses on
+_what_ to do and _why_, and only recommends optional resources about _how_ to do
+it. These practices:
+
+* are aspirational but when they can't be followed it's nice to explain why;
```
[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd]: This moved to README. Okay, let me spell out what I mean. Can you please help me by transforming these list into a phrase we can write in the doc?

* What: The aspirational practice that the reader fails to follow.
* Where: Somewhere. I guess the most common places would be the commit title, commit body, and PR comment.
* When: Whenever the reader fails to follow one of these practices.
* How: In writting.
* Who: The person who fails to follow one of these practices.

[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd]: so, something like....

These are aspirational goals, but we realize there may be times when you can not achieve all of them. In those cases, it's nice to let others know, e.g. in a commit description or a PR comment, how or why some of these guidelines could not be met.

[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd]: Thanks! c86b7d4 uses your suggestion.

File: git.Rmd (hunk 16 above)
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd]: I can't think of a rule of thumb for this, I guess that's why I brought it up. I'm trying to imagine the perspective when you're making a commit.... you make the changes, stage the changes, then write a commit message when you're committing it that says something like "this commit does XYZ". At that point, chances are that your commit actually does completely do XYZ, unless you accidentally did something unintentional, which almost by definition means you would not be consciously aware enough of it to apply any best practice guidance to avoid it.

So what's the intuition? I think maybe it's not about your commit doing what you said it does, which it probably does unless you made an unintentional mistake. It's about striving to combine like changes into a single commit so that all relevant changes related to a conceptual change are combined into one? So the guidance is something like "before you commit changes, consider if other changes you've made are part of the same conceptual change and combine them into one commit rather than committing them separately"?

[the remark this answers is not in the document]
Comment (maurolepore) [reply git.Rmd]: Thanks! The entire section now reads

> Each commit and feature should do what it promises, completely; and reverting it
should undo what that commit or feature meant to do, also completely. Before you
commit changes, consider if other changes you've made are part of the same
conceptual change and combine them into one commit rather than committing them
separately.

I removed everything else because it seems to be saying more or less the same. But let me know if you thing that some or all of that should stay.

Removed:

> If the goal of one commit requires changes in two files, that one commit should
include changes to both files. For example, if two commits in the same feature
branch promise to "Fix all typos", each commit is likely incomplete; they should
have been squashed into a single commit before sharing them. Similarly, if the
goal of a feature requires two kinds of changes, that feature should likely
include two commits.

[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd]: I like this much shorter version. I'm still not 100% convinced about what it says, but I think it's short and sweet enough to get the general concept across without confounding the topic so much that one is left like 🤔 

I'm also realizing how much my perspective on this and other sections here is impacted by my primary means of git usage (GUI client)... the concept of "staging" multiple and/or subsequent changes is so transparent in my workflow that it practically doesn't exist.

File: git.Rmd (hunk 18 above)
[the remark this answers is not in the document]
Comment (cjyetman) [reply git.Rmd]: so, I guess then my suggestion here would be... "Shared commits should never change." is not nearly specific or precise enough given that many people, myself 100% included, will not know what, if any, actions they intend to take will go against this guidance, i.e. "change a shared commit"

**Comment by MonikaFu** [comment]
Our Best Practices on working with git have been finalised!
